nginx/lua/init_worker.lua

```diff
@@ -0,0 +1,127 @@
+local config = require("config")
+local resty_http = require("resty.http")
+local jwt = require("resty.jwt")
+local cjson = require("cjson")
+local gossip = require("gossip")
+
+-- if file does not exist, we take the default values
+config.load("/etc/nginx/lua/config.json")
+
+local token_userpass = config.data.openidc_client_id .. ":" .. config.data.openidc_client_secret
+if token_userpass == ":" then
+    ngx.log(ngx.ERR, "No openidc_client_id or openidc_client_secret set in config.json")
+    return
+end
+
+-- This function is called by the gossip timer
+---@type function
+---@param peer string
+---@param message string
+---@param token string
+local function peer_exchange_gossip(peer, message, token)
+    local httpc = resty_http.new()
+    -- TODO: handle connection refused
+    local res, err = httpc:request_uri(peer .. "/gossip", {
+        method = "POST",
+        body = message,
+        headers = {
+            ["Content-Type"] = "application/json",
+            ["Content-Length"] = #message,
+            ["Authorization"] = "Bearer " .. token,
+            ["Accept"] = "application/json",
+        },
+    })
+    if not res then
+        local errmsg = "Failed to send gossip message to " .. peer .. ": " .. err
+        --gossip.handle_peer_error(peer, errmsg)
+        ngx.log(ngx.ERR, errmsg)
+    end
+    if res.status ~= 200 then
+        local errmsg = "Failed to send gossip message to " .. peer .. ": " .. res.status
+        ngx.log(ngx.ERR, errmsg)
+    end
+    if res.body == "" then
+        local errmsg = "Empty response from " .. peer
+        ngx.log(ngx.ERR, errmsg)
+    end
+    gossip.handle_message(res.body)
+end
+
+
+-- This function is called by the gossip timer
+---@type function
+---@param premature boolean
+---@return nil
+local function worker_gossip(premature)
+    if premature then
+        return
+    end
+
+    local tic = ngx.now()
+
+    -- Update our bearer token
+    local token = ngx.shared.gossip_data:get("bearer_token")
+    if type(token) ~= "string" then
+        token = nil
+    end
+    if not token or jwt:load_jwt(token).payload.exp < ngx.now() + 600 then
+        local httpc = resty_http.new()
+        local res = httpc:request_uri(config.data.openidc_iss .. "token", {
+            method = "POST",
+            -- TODO: make the scope configurable
+            body = "grant_type=client_credentials&scope=storage.read%3A%2F+hepcdn.access",
```

Comment:
The `hepcdn.access` scope is what we use to authenticate ourselves to our peers (we deserve to be in the federation) and the `storage.read:/` scope is so that we can execute `HEAD` requests to see if any of our peers have a file. (the HEAD requests will be implemented in a separate PR)